File: 04-hash-tables/03-hash-functions/cpp/hash_functions_demo.cpp
// 03 雜湊函數示範程式（C++）/ Hash functions demo program (C++).  // Bilingual file header.

#include "HashFunctions.hpp"  // Import hash functions and analysis helpers.
#include "UniversalHashing.hpp"  // Import universal hashing helpers.

#include <iostream>  // Provide std::cout for console output.
#include <vector>  // Provide std::vector for demo key lists.

int main() {  // Entry point for the demo CLI.
    int testKey = 123456;  // Pick a deterministic integer key for demo output.
    int mPrime = 97;  // Use a prime bucket count for division hashing.
    std::cout << "=== Hash Functions Demo (C++) ===\n";  // Print header.
    std::cout << "Key: " << testKey << ", m: " << mPrime << "\n";  // Print key summary.
    std::cout << "  Division:       " << hashfunctionsunit::divisionHash(testKey, mPrime) << "\n";  // Print division hash.
    std::cout << "  Multiplication: " << hashfunctionsunit::multiplicationHash(testKey, mPrime) << "\n";  // Print multiplication hash.
    std::cout << "  Mid-Square (4): " << hashfunctionsunit::midSquareHash(testKey, 4) << "\n";  // Print mid-square hash.
    std::cout << "  Folding (3):    " << hashfunctionsunit::foldingHash(testKey, 3, mPrime) << "\n";  // Print folding hash.

    std::vector<std::string> strings = {"hello", "world", "hash", "table", "cpp"};  // Sample strings.
    std::cout << "\n--- String hashes ---\n";  // Print section title.
    for (const std::string& s : strings) {  // Print a few hash values for each string.
        std::uint32_t djb2 = hashfunctionsunit::djb2Hash(s);  // Compute DJB2 hash.
        std::uint32_t fnv = hashfunctionsunit::fnv1aHash(s);  // Compute FNV-1a hash.
        std::uint32_t poly = hashfunctionsunit::polynomialHash(s, 31u, 1000003u);  // Compute polynomial hash mod a prime.
        std::cout << s << " => djb2=" << djb2 << ", fnv1a=" << fnv << ", poly=" << poly << "\n";  // Print one line.
    }  // Close loop.

    std::vector<std::string> keys;  // Deterministic keys list for distribution demo.
    for (int i = 0; i < 1000; i++) {  // Create 1000 keys.
        keys.push_back("key_" + std::to_string(i));  // Append one key.
    }  // Close loop.
    auto report = hashfunctionsunit::analyzeDistribution(  // Analyze distribution for DJB2 reduced mod 100.
        [](const std::string& k, int m) { return static_cast<int>(hashfunctionsunit::djb2Hash(k) % static_cast<std::uint32_t>(m)); },  // Hash adapter.
        keys, 100  // Provide keys and bucket count.
    );  // Close call.
    std::cout << "\n--- Distribution (DJB2, 1000 keys, m=100) ---\n";  // Print section title.
    std::cout << "non_empty=" << report.nonEmptyBuckets << ", max=" << report.maxBucketSize << ", std_dev=" << report.stdDeviation << "\n";  // Print summary.

    hashfunctionsunit::UniversalHashFamily uh(10, 123u);  // Universal hash family for ints with deterministic seed.
    std::cout << "\n--- UniversalHashFamily (m=10) ---\n";  // Print section title.
    std::cout << "hash(42)=" << uh.hash(42) << ", hash(137)=" << uh.hash(137) << "\n";  // Print example values.
    return 0;  // Exit success.
}  // Close main().

File: 04-hash-tables/03-hash-functions/javascript/hashFunctionsDemo.js
// 03 雜湊函數示範程式（JavaScript）/ Hash functions demo program (JavaScript).  // Bilingual file header.

"use strict";  // Enable strict mode for safer JavaScript semantics.

const {  // Import hash functions and analysis helper.  
    divisionHash,  // Division method hash.  
    multiplicationHash,  // Multiplication method hash.  
    midSquareHash,  // Mid-square hash.  
    foldingHash,  // Folding hash.  
    polynomialHash,  // Polynomial rolling hash.  
    djb2Hash,  // DJB2 hash.  
    fnv1aHash,  // FNV-1a hash.  
    analyzeDistribution,  // Distribution analyzer.  
} = require("./hashFunctions.js");  // Load module relative to this file.  

const {  // Import universal hash family for a tiny example.  
    UniversalHashFamily,  // Universal hash family for ints.  
} = require("./universalHashing.js");  // Load module relative to this file.  

function main() {  // CLI entry point for the demo.
    const testKey = 123456;  // Pick a deterministic integer key for demo output.
    const mPrime = 97;  // Use a prime bucket count for division hashing.
    console.log("=== Hash Functions Demo (JavaScript) ===");  // Print header.
    console.log(`Key: ${testKey}, m: ${mPrime}`);  // Print key summary.
    console.log(`  Division:       ${divisionHash(testKey, mPrime)}`);  // Print division hash.
    console.log(`  Multiplication: ${multiplicationHash(testKey, mPrime)}`);  // Print multiplication hash.
    console.log(`  Mid-Square (4): ${midSquareHash(testKey, 4)}`);  // Print mid-square hash.
    console.log(`  Folding (3):    ${foldingHash(testKey, 3, mPrime)}`);  // Print folding hash.

    console.log("");  // Blank line.
    console.log("--- String hashes ---");  // Print section title.
    for (const s of ["hello", "world", "hash", "table", "js"]) {  // Print a few hash values for each string.
        const djb2 = djb2Hash(s);  // Compute DJB2 hash.
        const fnv = fnv1aHash(s);  // Compute FNV-1a hash.
        const poly = polynomialHash(s, 31, 1000003);  // Compute polynomial hash mod a prime.
        console.log(`${s} => djb2=${djb2}, fnv1a=${fnv}, poly=${poly}`);  // Print one line.
    }  // Close loop.

    const keys = [];  // Deterministic keys list for distribution demo.
    for (let i = 0; i < 1000; i += 1) {  // Create 1000 keys.
        keys.push(`key_${i}`);  // Append one key string.
    }  // Close loop.
    const report = analyzeDistribution((k, m) => (djb2Hash(k) % m), keys, 100);  // Analyze DJB2 reduced mod 100.
    console.log("");  // Blank line.
    console.log("--- Distribution (DJB2, 1000 keys, m=100) ---");  // Print section title.
    console.log(`non_empty=${report.nonEmptyBuckets}, max=${report.maxBucketSize}, std_dev=${report.stdDeviation.toFixed(3)}`);  // Print summary.

    const uh = new UniversalHashFamily(10, null, 123);  // Universal hash family with deterministic seed.
    console.log("");  // Blank line.
    console.log("--- UniversalHashFamily (m=10) ---");  // Print section title.
    console.log(`hash(42)=${uh.hash(42)}, hash(137)=${uh.hash(137)}`);  // Print example values.
}  // End main().

if (require.main === module) {  // Execute demo only when run directly.
    main();  // Run main.
}  // Close direct-run guard.

